File: python/rewire_tracer/detect_ci.py
from __future__ import annotations

import os
from typing import NamedTuple


class CiContext(NamedTuple):
    platform: str
    run_id: str | None


def detect_ci() -> CiContext:
    if run_id := os.environ.get("GITHUB_RUN_ID"):
        return CiContext(platform="github_actions", run_id=run_id)
    if run_id := os.environ.get("CIRCLE_WORKFLOW_ID"):
        return CiContext(platform="circleci", run_id=run_id)
    if run_id := os.environ.get("CI_PIPELINE_ID"):
        return CiContext(platform="gitlab", run_id=run_id)
    if run_id := os.environ.get("REWIRE_RUN_ID"):
        return CiContext(platform="generic", run_id=run_id)
    return CiContext(platform="unknown", run_id=None)

File: ruby/lib/rewire/tracer/detect_ci.rb
# frozen_string_literal: true

module Rewire
  module Tracer
    CiContext = Struct.new(:platform, :run_id, keyword_init: true)

    def self.detect_ci
      if (run_id = ENV["GITHUB_RUN_ID"])
        CiContext.new(platform: "github_actions", run_id: run_id)
      elsif (run_id = ENV["CIRCLE_WORKFLOW_ID"])
        CiContext.new(platform: "circleci", run_id: run_id)
      elsif (run_id = ENV["CI_PIPELINE_ID"])
        CiContext.new(platform: "gitlab", run_id: run_id)
      elsif (run_id = ENV["REWIRE_RUN_ID"])
        CiContext.new(platform: "generic", run_id: run_id)
      else
        CiContext.new(platform: "unknown", run_id: nil)
      end
    end
  end
end
